### ISSUE TITLE
{{PLURAL}} is not always plural

### ISSUE THREAD

**Issue by desb42** [gnosygnu/xowa#633]
I have code that when displaying Category: pages shows the various totals in the local language by referencing the Mediawiki: pages

That is, for `fr.wikisource.org/wiki/Cat%C3%A9gorie:Religions_et_croyances`, the category totals template text will be found in `fr.wikisource.org/wiki/Mediawiki:Category-subcat-count`

Whereas xowa uses the info from (in this case) bin\any\xowa\cfg\lang\core\fr.gfs

In fr.gfs the template text is:

    Cette catégorie comprend {{PLURAL:~{1}|la sous-catégorie|~{1} sous-catégories, dont {{PLURAL:~{0}|celle|les ~{0}}}}} ci-dessous.

from mediawiki (with adjustment) the text is:

    Cette catégorie {{PLURAL:~{1}|0=ne comprend aucune sous-catégorie|1=comprend seulement la sous-catégorie ci-dessous|comprend ~{1} sous-catégories, dont {{PLURAL:~{0}|0=aucune|1=celle|les ~{0}}} ci-dessous}}.

It seems PLURAL has gained a feature I was not aware of
Taking the last plural example (as it is short)

    {{PLURAL:~{0}|0=aucune|1=celle|les ~{0}}}
There are specific tests for a value (in this case '0' and '1' and then the default) acting a bit like a #switch 

If I substitute '15' for '~{0}', the text is

    {{PLURAL:15|0=aucune|1=celle|les 15}}
and the result in xowa is:

    1=celle

**Comment by gnosygnu** [comment]
First, thanks as always for the detailed breakdown. I know it takes a while to track these odd behaviors down, and it's super-helpful to have the guided walk-through.

I committed the main part above. I'm going to do another commit to clean up some stuff before moving this over to DONE. Then I'll start working on some of the other issues from the past weeks.

Some design notes below. Thanks!

---

For this issue, I ended up taking the longer route because I wanted to match the MediaWiki plurals code as close as possible.

The benefit of this approach is that the XOWA Java code now is a pretty close representation of the MediaWiki PHP code. Any behaviors (and associated bugs) should be reproduced faithfully. For example:

* MediaWiki PHP: https://github.com/wikimedia/mediawiki-vendor/blob/master/wikimedia/cldr-plural-rule-parser/src/Converter.php
* XOWA Java: https://github.com/gnosygnu/xowa/blob/master/400_xowa/src/gplx/xowa/mediawiki/vendor/wikimedia/cldr_plural_rule_parser/src/XomwConverter.java

The drawback is that it takes a little more time to code, especially to simulate some of the underlying PHP behaviors. This should get easier as time goes on, but this is a growing pain.

Finally, there is another potential lurking drawback in terms of performance. I feel that MediaWiki is not always written in a performant way, and this will get exacerbated by the pseudo-PHP'ization by Java. 

In general, MediaWiki can get away with slower code as it amortizes any performance cost by incrementally parsing and caching over its large user-base and wide cache-coverage. As an extreme example, a page can take 1 minute to parse, but this 1 minute cost is only borne by the editor, and every viewer afterwards gets to see the page in a fraction of a second. In contrast, XOWA is basically single-user, and the HTML generation is basically a single-use scenario (take this page; parse it; compile it; repeat for the other 5 million pages).

For something like the plural-rules, I'm not too worried about slow performance, but as I transcribe more and more MediaWiki code, the aggregate costs will pile up. I have some strategies on identifying these bottlenecks, but I am dreading the time it takes to do the first full HTML-database generation with a MediaWiki-like parser (whenever XOWA gets there)
